File: Exavision.Seasense.Shared/Capabilities/Camera/CameraZoomAbsolutePositionCapability.cs
using Exavision.Seasense.Shared.Settings;
using Exavision.Seasense.Shared.States;
using Exavision.Seasense.Shared.States.Camera;

namespace Exavision.Seasense.Shared.Capabilities.Camera {
    public abstract class CameraZoomAbsolutePositionCapability<S> : Capability<S>, ICameraZoomAbsolutePositionCapability where S : SettingCapability, new() {
        public abstract double HorizontalFieldOfView { get; }

        public abstract double ZoomMultiplier { get; }

        public override S GetSetting(S setting) {
            setting.CapabilityType = CapabilityType.CameraZoomAbsolutePosition;
            return setting;
        }
        public override CapabilityState GetState() {
            CameraZoomAbsolutePositionState state = new CameraZoomAbsolutePositionState() {
                Id = this.Id,
                HorizontalFieldOfView = this.HorizontalFieldOfView,
                ZoomMultiplier = this.ZoomMultiplier
            };
            return state;
        }
    }
}

File: Exavision.Seasense/Client/src/app/materials/capabilities/camera/camera-zoom-absolute-position-capability.ts
import { Material } from "../../material";
import { SettingCapability } from "../../settings/setting-capability";
import { Site } from "../../site";
import { CameraZoomAbsolutePositionStep } from "../../states/camera/camera-zoom-absolute-position-state";
import { CapabilityState } from "../../states/capability-state";
import { Unit } from "../../unit";
import { Capability } from "../capability";
import { CapabilityType } from "../capability-type";

export class CameraZoomAbsolutePositionCapability implements Capability {
 
  id: string;
  capabilityType: CapabilityType;
  material: Material | undefined;
  zoomMultiplier: number = 1;
  horizontalFieldOfView: number = 25;


  constructor(settingCapability: SettingCapability) {
    this.id = settingCapability.id;
    this.capabilityType = CapabilityType.CameraZoomAbsolutePosition;
  }

 
  setState(valueState: CapabilityState): void {
    let state: CameraZoomAbsolutePositionStep = <CameraZoomAbsolutePositionStep>valueState;
    this.zoomMultiplier = state.zoomMultiplier;
    this.horizontalFieldOfView = state.horizontalFieldOfView;
  }
}
